### ISSUE TITLE
Extending QUIC notation 

### ISSUE THREAD

**Issue by LPardue** [moq-wg/moq-transport#536]
In https://moq-wg.github.io/moq-transport/draft-ietf-moq-transport.html#section-1.3 three new notations are presented thst extend the usual QUIC notations defined in RFC 9000.

Personally, with no hats, I don't see much value in adding all the notation. It introduces overlap and divergence from how we already use the existing notation to convey similar types e.g. the `x (b)` form

> SUBSCRIBE_NAMESPACE_ERROR
{
  Track Namespace Prefix (tuple),
  Error Code (i),
  Reason Phrase (b),
}

vs two separate fields

> CONNECTION_CLOSE Frame {
  Type (i) = 0x1c..0x1d,
  Error Code (i),
  [Frame Type (i)],
  Reason Phrase Length (i),
  Reason Phrase (..),
}

The `x (tuple)`  format seems potentially useful but grates me a bit given its just a group of things, where the group size is restricted to 1..32. It could be represented for example as a type

``` 
Track Namespace Parameters {
  Parameter Count (8) = 1..32,
  Track Namespace Parameter (..) ...,
}

Track Namespace Parameter {
  Track Namespace Paramater Length (i),
  Track Namespace Parameter Value (..),
}
```

Personally, I don't quite think the value of the new things overcomes the additional cognitive load that readers need to carry. There's many other folks that already use or consume RFC 9000 notation. This issue is sort of already present in MoQT, where the `Parameter` type doesn't use the `x (b)` form even though it could to cover the length and value parts.

```
Parameter {
  Parameter Type (i),
  Parameter Length (i),
  Parameter Value (..),
}
```

So one suggestion is to remove the special notations and go back to RFC 9000 style. This shouldn't change the wire encoding at all.

A different suggeation, if I'm in the rough and the MoQ WG has consensus that there is value in extending the notation and dealing with the overlap and cognitive load, it's possible other WGs might too. Having to import MoQT for a notational convention is messy. So it might make sense to spin them out into a separate I-D that could update RFC 9000, and send that over to the QUIC WG to manage.

**Comment by fluffy** [comment]
I don't really care what we do here either but I would like to point at what I suspect is one of the root causes of all the pain around this is given 

```
Track {
   Len (i), 
   Val (..),
  } 
  ''' 
  
It is not possible to write a encoder/parser for that because the length of val may or may not be encoded in Len. You have to have extra information to know how to write an parser for that structure. That seems like a pretty big lacking in any formal syntax.

**Comment by LPardue** [comment]
I agree with your point. However, knowing what exactly to do also depends on the context of the parsing. There could be redundant length fields etc.

And there's formal syntax and there's formal syntax :-) despite what our friends in the formal methods world might wish for, I don't think anyone wants to devolve into Z notation etc. 

We've done pretty ok with the existing notation plus normative language across a number of RFCs. It's flexible while providing a natural mapping to many common languages. That might be good enough, or there might be an opportunity to take feedback to the QUIC WG and let them consider it. Adding notation inside MoQT seems like the least best outcome to me

**Comment by suhasHere** [comment]
No strong opinion. Other than (b) which is a simplification, the draft currently defines 2 new notations
1. Boolean
2. Tuple
I feel those 2 are still needed as RFC9000 doesn't really define them.

**Comment by LPardue** [comment]
The boolean `x (f)` is used in two places, for example

> SUBSCRIBE_OK
{
  Subscribe ID (i),
  Expires (i),
  Group Order (8),
  ContentExists (f),
  [Largest Group ID (i)],
  [Largest Object ID (i)],
  Number of Parameters (i),
  Subscribe Parameters (..) ...
}

This especially sticks out because the `Group Order` field is also a bit field but it uses a different value range.

> Group Order: Indicates the subscription will be delivered in Ascending (0x1) or Descending (0x2) order by group. See [Section 4](https://moq-wg.github.io/moq-transport/draft-ietf-moq-transport.html#priorities). Values of 0x0 and those larger than 0x2 are a protocol error.

> ContentExists: 1 if an object has been published on this track, 0 if not. If 0, then the Largest Group ID and Largest Object ID fields will not be present.

You can substitute the `x (f)` notation e.g. `contentExists(8),` and add the statement `values other than 0x0 or 0x1 are a protocol error`.

The other common way of defining message variants, that may or may not contain an optional field, is to use the type value itself. For example, 

> CONNECTION_CLOSE Frame {
  Type (i) = 0x1c..0x1d,
  Error Code (i),
  [Frame Type (i)],
  Reason Phrase Length (i),
  Reason Phrase (..),
}

> Frame Type: A variable-length integer encoding the type of frame that triggered the error. A value of 0 (equivalent to the mention of the PADDING frame) is used when the frame type is unknown. The application-specific variant of CONNECTION_CLOSE (type 0x1d) does not include this field.

This approach is used in several RFCs. It avoids needing to send an entire byte for the field.

As noted above, the tuple type can already be represented with existing notation. As I read more, I think the notation actually causes a disservice to implementers because the tuple size is represented as a varint but MoQ allows only values in the range 1-32. That constraint is stated once in https://moq-wg.github.io/moq-transport/draft-ietf-moq-transport.html#section-2.4.1. There is no language to state what happens when other values are received. By creating a generic notation, MoQT actually made things harder to decipher.

**Comment by suhasHere** [comment]
> The boolean `x (f)` is used in two places, for example
> 
> > SUBSCRIBE_OK
> > {
> > Subscribe ID (i),
> > Expires (i),
> > Group Order (8),
> > ContentExists (f),
> > [Largest Group ID (i)],
> > [Largest Object ID (i)],
> > Number of Parameters (i),
> > Subscribe Parameters (..) ...
> > }
> 
> This especially sticks out because the `Group Order` field is also a bit field but it uses a different value range.
> 
> > Group Order: Indicates the subscription will be delivered in Ascending (0x1) or Descending (0x2) order by group. See [Section 4](https://moq-wg.github.io/moq-transport/draft-ietf-moq-transport.html#priorities). Values of 0x0 and those larger than 0x2 are a protocol error.
> 
> > ContentExists: 1 if an object has been published on this track, 0 if not. If 0, then the Largest Group ID and Largest Object ID fields will not be present.
> 
> You can substitute the `x (f)` notation e.g. `contentExists(8),` and add the statement `values other than 0x0 or 0x1 are a protocol error`.
> 

[Suhas] That exactly how it was defined to begin with. Just avoids repeating the similar things in multiple places and then one. I do understand it is just 2 places. I wish QUIC had defined common types (including boolean :-) ) 

> The other common way of defining message variants, that may or may not contain an optional field, is to use the type value itself. For example,
> 
> > CONNECTION_CLOSE Frame {
> > Type (i) = 0x1c..0x1d,
> > Error Code (i),
> > [Frame Type (i)],
> > Reason Phrase Length (i),
> > Reason Phrase (..),
> > }
> 
> > Frame Type: A variable-length integer encoding the type of frame that triggered the error. A value of 0 (equivalent to the mention of the PADDING frame) is used when the frame type is unknown. The application-specific variant of CONNECTION_CLOSE (type 0x1d) does not include this field.
> 
> This approach is used in several RFCs. It avoids needing to send an entire byte for the field.
> 
> As noted above, the tuple type can already be represented with existing notation. As I read more, I think the notation actually causes a disservice to implementers because the tuple size is represented as a varint but MoQ allows only values in the range 1-32. That constraint is stated once in https://moq-wg.github.io/moq-transport/draft-ietf-moq-transport.html#section-2.4.1. There is no language to state what happens when other values are received. By creating a generic notation, MoQT actually made things harder to decipher.

**Comment by afrind** [comment]
See #1192 which removes x (tuple) and x (b).
